torch/distributions/transforms.py

```diff
@@ -0,0 +1,342 @@
+    :meth:`_inverse`. Derived classes that set `bijective=True` should also
+    implement :meth:`log_abs_det_jacobian`.
+    """
+    bijective = False
+
+    def __init__(self, cache_size=0):
+        if cache_size != 0:
+            if cache_size == 1:
+                self._cached_x_y = None, None
+                self.__call__ = self._cached_call
+                self.inverse = self._cached_inverse
+            else:
+                raise NotImplementedError('cache_size must be 0 or 1')
+
+    @lazy_property
+    def inv(self):
+        """
+        Returns the inverse :class:`Transform` of this transform.
+        """
+        return InverseTransform(self)
+
+    def __eq__(self, other):
+        return self is other
+
+    def __ne__(self, other):
+        # Necessary for Python2
+        return not self.__eq__(other)
+
+    def __call__(self, x):
+        """
+        Computes the transform `x => y`.
+        """
+        return self._call(x)
+
+    def inverse(self, y):
+        """
+        Inverts the transform `y => x`.
+        """
+        return self._inverse(y)
+
+    def _cached_call(self, x):
+        """
+        Computes the memoized transform `x => y`.
+        """
+        x_old, y_old = self._cached_x_y
+        if x is x_old:
+            return y_old
+        y = self._call(x)
+        self._cached_x_y = x, y
+        return y
+
+    def _cached_inverse(self, y):
+        """
+        Inverts the memoized transform `y => x`.
+        """
+        x_old, y_old = self._cached_x_y
+        if y is y_old:
+            return x_old
+        x = self._inverse(y)
+        self._cached_x_y = x, y
+        return x
+
+    def _call(self, x):
+        """
+        Abstract method to compute forward transformation.
+        """
+        raise NotImplementedError
+
+    def _inverse(self, y):
+        """
+        Abstract method to compute inverse transformation.
+        """
+        raise NotImplementedError
+
+    def log_abs_det_jacobian(self, x, y):
+        """
+        Computes the log det jacobian `log |dy/dx|` given input and output.
+        """
+        raise NotImplementedError
+
+
+class InverseTransform(Transform):
+    """
+    Inverts a single :class:`Transform`.
+    """
+    __slots__ = ['inv']
+
+    def __init__(self, transform, cache_size=0):
+        self.inv = transform
+        super(InverseTransform, self).__init__(cache_size=cache_size)
+
+    @constraints.dependent_property
+    def domain(self):
+        return self.inv.codomain
+
+    @constraints.dependent_property
+    def codomain(self):
+        return self.inv.domain
+
+    @property
+    def bijective(self):
+        return self.inv.bijective
+
+    def __eq__(self, other):
+        if not isinstance(other, InverseTransform):
+            return False
+        return self.inv == other.inv
+
+    def _call(self, x):
+        return self.inv.inverse(x)
+
+    def _inverse(self, y):
+        return self.inv.__call__(y)

```

Comment:
Shouldn't this call `_call` (since `__call__` might do some extra things like caching)?

Reply:
As per your comment above, I was able to remove `InverseTransform._call` and `InverseTransform._inverse` in favor of the single delegating `InverseTransform.__call__`. Because this is different behavior than other `Transform`s, I've also made `_InverseTransform` private to prevent a user from accidentally creating `InverseTransform(InverseTransform(t))`. All inversion should now happen through the `.inv` property.